hermes/probe/monitor.go

```diff
@@ -12,9 +12,160 @@
+type Target struct {
+	// Target stores the proto config for the target to be probed.
+	Target *probepb.Target
+
+	// Journal stores the state of MonitorProbe as a combination of a next operation intent enum and a filenames map.
+	Journal *journalpb.StateJournal
+
+	// LatencyMetrics stores the api call and probe operation latency for a given target run.
+	// Metrics are stored with additional labels to record operation type and exit status.
+	LatencyMetrics *metrics.Metrics
+}
 
 // MonitorProbe holds aggregate information about all probe runs, per-target.
 // It also holds the config and options used to initialise the probe.
-type MonitorProbe struct{}
+type MonitorProbe struct {
+	name    string
+	config  *probepb.HermesProbeDef
+	targets []*Target
+	opts    *options.Options
+	logger  *logger.Logger
+}
+
+// Init initializes the probe with the given parameters.
+// This is a required method to implement the cloudprober.Probes.Probe interface.
+func (mp *MonitorProbe) Init(name string, opts *options.Options) error {
+	// Cast from Cloudprober defined interface to HermesProbeDef.
+	// This allows for accessing the variables and methods of a HermesProbeDef object.
+	conf, ok := opts.ProbeConf.(*probepb.HermesProbeDef)
+	if !ok {
+		return fmt.Errorf("invalid argument: opts.ProbeConf is not of type *probepb.HermesProbeDef")
+	}
+	mp.name = name
+	mp.config = conf
+
+	mp.targets = make([]*Target, 0)
+	for _, t := range mp.config.GetTargets() {
+		run := &Target{
+			Target: t,
+			Journal: &journalpb.StateJournal{
+				Intent:    &journalpb.Intent{},
+				Filenames: make(map[int32]string),
+			},
+		}
+		mp.targets = append(mp.targets, run)
+	}
+	mp.opts = opts
+	mp.logger = opts.Logger
+
+	return nil
+}
+
+// Start runs the probe indefinitely, unless cancelled, at the configured interval.
+// Probe metrics will be sent via the metricChan at the end of the probe run.
+// This is a required method to implement the cloudprober.Probes.Probe interface.
+// Arguments:
+//	- ctx: context provided for cancelling probe.
+//	- metricChan: unidirectional channel used for sending metrics to be surfaced.
+func (mp *MonitorProbe) Start(ctx context.Context, metricChan chan<- *cpmetrics.EventMetrics) {
+	probeTicker := time.NewTicker(mp.opts.Interval)
+
+	for {
+		select {
+		case <-ctx.Done():
+			probeTicker.Stop()
+			return
+		case <-probeTicker.C:
+			mp.runProbe(ctx, metricChan)
+		}
+	}
+}
+
+// reportMetrics sends the metrics recorded in the current probe run to Cloudprober.
+// Arguments:
+//	- run: metrics from a probe run on a target.
+//	- metricChan: metric channel passed from Cloudprober.
+func reportMetrics(run *metrics.Metrics, metricChan chan<- *cpmetrics.EventMetrics) {
+	for _, op := range run.ProbeOpLatency {
+		for _, m := range op {
+			metricChan <- m
+		}
+	}
+
+	for _, call := range run.ApiCallLatency {
+		for _, m := range call {
+			metricChan <- m
+		}
+	}
+}
+
+// runProbe runs probe for all targets and update EventMetrics.
```

Comment:
Update this comment.

Reply:
Updated to remove unexported symbol mention and to be more descriptive.